pkgs/_macro_builder/lib/macro_builder.dart

```diff
@@ -2,22 +2,110 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+import 'dart:io';
+
 import 'package:dart_model/dart_model.dart';
 
+/// Builds macros.
+///
+/// TODO(davidmorgan): add a way to clean up generated files and built output.
 class MacroBuilder {
   /// Builds an executable from user-written macro code.
   ///
   /// Each `QualifiedName` in [macroImplementations] must point to a class that
   /// implements `Macro` from `package:macro`.
+  ///
+  /// The [packageConfig] must include the macros and all their deps.
   Future<BuiltMacroBundle> build(
-      Iterable<QualifiedName> macroImplementations) async {
-    // TODO(davidmorgan): implement.
-    // Generated entrypoint will instantiate all the `Macro` instances pointed
-    // to by `macroImplementations` then pass them to `MacroClient.run` in
-    // `package:_macro_client`.
-    return BuiltMacroBundle();
+      File packageConfig, Iterable<QualifiedName> macroImplementations) async {
+    final script = createBootstrap(macroImplementations.toList());
+
+    return await MacroBuild(packageConfig, script).build();
+  }
+
+  /// Creates the entrypoint script for [macros].
+  static String createBootstrap(List<QualifiedName> macros) {
```

Comment:
For the "private" packages, starting with an underscore, I put implementation directly under `lib` instead of `lib/src`; my expectation is that we end up folding them entirely into some package's `src` so adding `lib/src` now just seemed like noise.

So I'm thinking of this entire file as private.

Happy to do something else though, as you like.

Reply:
I do think there is a difference here worth encapsulating.

If I understand correctly, MacroBuilder will be public (ie: accessible to the CFE/Analyzer). But this function is not intended to be used by either, and should ideally not be exposed.

Reply:
That makes sense, moved out to `lib/src` as you originally suggested. Thanks :)